sys/kern/kern_exec.c

```diff
@@ -1646,30 +1695,16 @@ exec_copyout_strings(struct image_params *imgp, uintcap_t *stack_base)
 	p = imgp->proc;
 	szsigcode = 0;
 
+	strings_on_stack = true;
 #if __has_feature(capabilities)
-	/*
-	 * destp must cover all of the smaller buffers copied out
-	 * onto the stack, so roundup the length to be representable
-	 * even if it means the capability extends beyond the stack
-	 * bounds.
-	 */
-	stack_vaddr = (vaddr_t)p->p_vmspace->vm_maxsaddr;
-	ssiz = p->p_sysent->sv_usrstack - stack_vaddr;
-	stack_offset = 0;
-	do {
-		rounded_stack_vaddr = CHERI_REPRESENTABLE_BASE(stack_vaddr,
-		    ssiz + stack_offset);
-		stack_offset = stack_vaddr - rounded_stack_vaddr;
-	} while (rounded_stack_vaddr != CHERI_REPRESENTABLE_BASE(stack_vaddr,
-	    ssiz + stack_offset));
-	destp = (uintcap_t)cheri_capability_build_user_data(
-	    CHERI_CAP_USER_DATA_PERMS, rounded_stack_vaddr,
-	    CHERI_REPRESENTABLE_LENGTH(ssiz + stack_offset), stack_offset);
-	destp = cheri_setaddress(destp, p->p_sysent->sv_psstrings);
+	if (imgp->stack != imgp->strings)
+		strings_on_stack = false;
+	destp = (uintcap_t)imgp->strings;
+	destp = cheri_setaddress(destp, p->p_psstrings);
 	arginfo = (struct ps_strings * __capability)cheri_setboundsexact(destp,
 	    sizeof(*arginfo));
 #else
-	arginfo = (struct ps_strings *)p->p_sysent->sv_psstrings;
+	arginfo = (struct ps_strings *)p->p_psstrings;
```

Comment:
Can we reorder this so the CHERI and non-CHERI kernel paths are in the same order?